criu/pidfd.c

```diff
@@ -0,0 +1,121 @@
+#include "imgset.h"
+#include "pidfd.h"
+#include "fdinfo.h"
+#include "protobuf.h"
+#include "pstree.h"
+
+#undef LOG_PREFIX
+#define LOG_PREFIX "pidfd: "
+
+struct pidfd_info {
+	PidfdEntry *pidfe;
+	struct file_desc d;
+};
+
+int is_pidfd_link(char *link)
+{
+	/*
+	* pidfs was introduced in Linux 6.9
+	* before which anonymous-inodes were used
+	*/
+	return is_anon_link_type(link, "[pidfd]");
+}
+
+static void pr_info_pidfd(char *action, PidfdEntry *pidfe)
+{
+	pr_info("%s: id %#08x flags %u mnt_id %d ino %u NSpid %u\n",
+		action, pidfe->id, pidfe->flags, pidfe->mnt_id,
+		pidfe->ino, pidfe->nspid
+	);
+}
+
+static int dump_one_pidfd(int pidfd, u32 id, const struct fd_parms *p)
+{
+	PidfdEntry pidfe = PIDFD_ENTRY__INIT;
+	FileEntry fe = FILE_ENTRY__INIT;
+
+	if (parse_fdinfo(pidfd, FD_TYPES__PIDFD, &pidfe))
+		return -1;
+
+	/*
+	* TODO:
+	* First we should check if the process is alive or not
+	* If it is dead:
+	*	dump a ghost pidfd
+	*/
+	/*
+	* Check if the pid pidfd refers to is part of process tree
+	* This ensures the process will exist on restore.
+	*/
+	if (!pstree_item_by_real(pidfe.pid)) {
+		pr_perror("pidfd pid %d is not a part of process tree..", pidfe.pid);
```

Comment:
I think we need to think about how to support pidfd of a dead process.

Reply:
I have added another test that checks for dead process support, and I will work on this over the next week.

Reply:
Hello Andrei!
After a discussion with Alex, I have updated the PR with dead process support.
Here's what I tried to do:
1. While collecting, create a hash table containing all dead pidfds, which stores the inode number and the number of times this inode occurred (count).
2. For each unique inode, we create a temporary process, open a pidfd to this temporary process and reduce the count. When this count reaches zero, we kill this temporary process and free allocated resources.

The hash table implementation tries to follow the implementation of `sk_info_hash` https://github.com/checkpoint-restore/criu/blob/criu-dev/criu/sk-unix.c#L1006

and `pidfd_hash` https://github.com/checkpoint-restore/criu/blob/93746eb25014f21287f2edc5f5dd7545ce19ab52/criu/pidfd-store.c#L26

I have also updated the `pidfd_dead` test so that it opens more that just one pair of pidfds.